### PR TITLE
feature/4-request-reply-adapter

### LINKED ISSUE #4: RequestReply adapter
When media was processed in core module then it should be redirected to adapter defined in HTTP path param. This task goal is to create the first, simple adapter using standard HTTP request-reply flow.

At this moment we do not have any adapter so whole background should be implemented to allow to use them in the future - smoothly.

I was thinking about interface like below.

```typescript
type BasicAdapterCommand<T> = {
	outputBuffer: Buffer, // Output buffer of processed media file
	fileType: 'image' | 'video', // Info if processed media file is image or video
	requestBody: T, // Request body - important it the future to process extra parameters (e.g in S3 integration)
};

type AdapterResult = { info?: any, buffer?: Buffer }; // When buffer property is set then HTTP response should return file, otherwise info object will be sent

interface Adapter<T> {
	resize(command: BasicAdapterCommand<T>): Promise<AdapterResult>;
	crop(command: BasidAdapterCommand<T>): Promise<AdapterResult>;
}
```

Every adapter should implement `Adapter` interface.

Remember about tests!!

To understand the whole flow better I attached block diagram.
image-service

### PR DESCRIPTION
**Pull request by i-plis**

### CODE REVIEW

File: src/adapters/adapter.types.ts [hunk 1]
```diff
@@ -8,6 +8,13 @@ export const ADAPTERS = Object.values(AdapterEnum);
 
 export type AdapterType = typeof AdapterEnum[keyof typeof AdapterEnum]
 
-export type Adapter = {
-  handleFile: (file: Stream) => unknown;
+export type BasicAdapterCommand<T> = {
+  file: Stream,
+  fileType: 'image' | 'video', // Info if processed media file is image or video
```
Comment (JonaszJestem) [inline src/adapters/adapter.types.ts]:
```suggestion
  fileType: 'image' | 'video',
```
imo comment is not needed

File: src/core/common/response.handler.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import type { AdapterResult } from '../../adapters/adapter.types';
+import type { FastifyReply } from 'fastify';
+import Boom from 'boom';
+
+export const handleResponse = async (adapterResult: AdapterResult, reply: FastifyReply): Promise<void> => {
+  if (adapterResult.file) {
+    reply
+      // .type('image/jpg')
```
Comment (JonaszJestem) [inline src/core/common/response.handler.ts]:
```suggestion

```

**Review by JonaszJestem — APPROVED** [review]

File: src/adapters/adapter.types.ts (hunk 1 above)
Comment (KrzysiekWyka) [reply src/adapters/adapter.types.ts]: Can u create `FileType` enum and use it here?

File: src/core/common/response.handler.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      .send(adapterResult.file)
+    ;
+    return;
+  }
+  // sending other info not implemented yet
```
Comment (KrzysiekWyka) [inline src/core/common/response.handler.ts]:
```suggestion
  // TODO: sending other info not implemented yet
```

**Review by KrzysiekWyka — CHANGES_REQUESTED** [review]

**Comment by soanvig** [comment]
Conflicts

File: src/adapters/adapter.utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { getAdapter } from './adapter.utils';
+import { AdapterEnum } from './adapter.types';
+import { requestReplyAdapter } from './requestReply.adapter';
+import { AdapterErrors } from './adapter.errors';
+
+describe('AdapterUtils', () => {
+  describe('getAdapter', () => {
+    it('should throw error', async () => {
+      expect(() => getAdapter('NonexistentAdapter')).toThrowError(AdapterErrors.AdapterNotFound);
```
Comment (KrzysiekWyka) [inline src/adapters/adapter.utils.test.ts:9]:
```suggestion
      await expect(getAdapter('NonexistentAdapter')).rejects.toThrowError(AdapterErrors.AdapterNotFound);
```

File: src/adapters/requestReply.adapter.test.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  getTestImage,
+} from '../../test/test.utils';
+import { FileTypeEnum } from './adapter.types';
+
+describe('RequestReplyAdapter', () => {
+  describe('handleFile', () => {
+    it('should return file', async () => {
+      const image = await getTestImage();
+
+      const adapterResult = await requestReplyAdapter.handleFile({ file: image, fileType: FileTypeEnum.image, requestBody: {}});
```
Comment (KrzysiekWyka) [inline src/adapters/requestReply.adapter.test.ts:12]: Missing test for video file type.

**Review by KrzysiekWyka — CHANGES_REQUESTED** [review]

File: src/adapters/adapter.utils.test.ts (hunk 4 above)
Comment (i-plis) [reply src/adapters/adapter.utils.test.ts:9]: I've tried your suggestion, but it fails... I'm not sure why...

File: src/adapters/requestReply.adapter.test.ts (hunk 5 above)
Comment (i-plis) [reply src/adapters/requestReply.adapter.test.ts:12]: Added

**Review by soanvig — APPROVED** [review]

**Review by KrzysiekWyka — APPROVED** [review]

**Pull request merged**